conans/test/integration/conanfile/test_deploy_method.py

```diff
@@ -0,0 +1,46 @@
+import textwrap
+
+from conans.test.utils.tools import TestClient
+
+
+def test_deploy_method():
+    c = TestClient()
+    conanfile = textwrap.dedent("""
+        import os
+        from conan import ConanFile
+        from conan.tools.files import copy, save
+        class Pkg(ConanFile):
+            name = "{name}"
+            version = "0.1"
+            {requires}
+            def package(self):
+                save(self, os.path.join(self.package_folder, f"my{name}file.txt"), "HELLO!!!!")
+            def deploy(self):
+                copy(self, "*", src=self.package_folder, dst=self.deploy_folder)
+            """)
+    c.save({"dep/conanfile.py": conanfile.format(name="dep", requires=""),
+            "pkg/conanfile.py": conanfile.format(name="pkg", requires="requires='dep/0.1'")})
+    c.run("create dep")
+    assert "Executing deploy()" not in c.out
+    c.run("create pkg")
+    assert "Executing deploy()" not in c.out
+
+    # Doesn't install by default
+    c.run("install --requires=pkg/0.1")
+    assert "Executing deploy()" not in c.out
+
+    # Doesn't install with other patterns
+    c.run("install --requires=pkg/0.1 --deployer-package=other")
+    assert "Executing deploy()" not in c.out
+
+    # install can deploy all
+    c.run("install --requires=pkg/0.1 --deployer-package=* --deployer-folder=mydeploy")
```

Comment:
Yep, I think that this is a great interface. You're right, we might want to deploy multiple packages at once. I'm wondering about overloading the `--deployer-folder` though - is the design intention here that each deployer is responsible for namespacing itself within the `deployer-folder`? For instance, if I run

```
$ conan install --requires=foo/0.1.0 --deployer-package=foo/0.1.0 --deployer=licenses --deployer=full_deploy --deployer-folder=/tmp/deploy-test
```
then each of the deployers (kind of) conflict:
```
$tree /tmp/deploy-test -L 2
/tmp/deploy-test
├── conaninfo.txt
├── conanmanifest.txt
├── full_deploy
│   └── host
├── include
│   └── foo.h
├── lib
│   └── libfoo.a
└── licenses.zip
```

`full_deploy` namespaces itself properly, licenses.py (from the conan-extensions repo) places licenses.zip in the root of the deployer folder, and in this case I've emulated your `deploy()` method, which copies the package folder to the root. If this is the most desirable design, then a best practice would need to emerge to avoid contaminating other deployments (i.e., by outputing to a subfolder of `deploy-folder`). Or is there another design that might remove the need for a best practice?

I guess removing the need for a best practice also has issues - this is the most flexible, and gives the user the most room to construct things the way they want, at the cost of a (potential) maintenance overhead in ensuring deployers deploy consistently if there's a possibility of a clash.

Reply:
Yes, I think it should be the responsibility of the users to namespace their deploys if that is what they want/need. If we try to namespace something, then it would be certainly blocking for users that want to deploy all .dlls in a single common folder to all dependencies for example, and they want to do it in their ``deploy()`` method.

Once deployers are running from different origins and different packages ``deploy()`` now, the space for conflict is always there (they could be deploying directly to ``/usr/bin`` etc, and conflicting anyway). The ``full_deploy`` implement the namespace in itself, it is not something external, with the intention that users can take it as inspiration and easily customize the namespace/layout of the deployment to fit their needs.

The rule of thumb is that deployers copy to user space, and as such it is the full responsibiliy of the user (via the different deployers implementations) to control that user space. And then, yes, some common practices might happen with some common namespaces patterns, but that should be it, just some common practices, not something mandated or implemented by the tool.